Выполнить ТЗ №10

SergeyLeshchev opened the pull request.

SergeyLeshchev committed:
Выполнить ТЗ №10

SergeyLeshchev committed:
Выполнить ТЗ №10

1. Внести правки кодстайл

SergeyLeshchev committed:
Выполнить ТЗ №10

1. Добавить поле rate в класс Film

Ivanov-s-tmn commented on pom.xml:
```diff
@@ -32,6 +35,30 @@
+		<dependency>
+			<groupId>ch.qos.logback</groupId>
+			<artifactId>logback-classic</artifactId>
+			<version>1.4.8</version>
```
Общие версии зависимостей хорошая практика выносить в блок properties, чтобы не дублировать, по аналогии сделано у спринг бут стартеров, но там немножко иной подход с родительским стартером. Плюс не очень понятна эта зависимость, если используется slf4j

SergeyLeshchev replied:
Эту зависимость я добавил, когда добавлял jackson в проект. Думал, что нужны обе зависимости, сейчас удалил и убедился, что работает и одна. Теперь я так понимаю и версию можно не убирать в properties

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,42 @@
+        this.description = description;
+        this.releaseDate = releaseDate;
+        this.duration = duration;
+        this.rate = rate;
```
Конструктор можно сгенерировать аннотацией `@AllArgsConstructor`

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,42 @@
+@NoArgsConstructor
 public class Film {
+    private Long id;
+    @NonNull
+    private String name;
+    @NonNull
+    private String description;
+    @NonNull
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
+    private LocalDate releaseDate;
+    @NonNull
```
Не вижу особого смысла в аннотации `@NonNull`, это аннотация ломбок, которая генерирует проверку на null и выбрасывает NPE. Имеет смысл тогда сразу реализовать валидация через javax.validation, и проверять поля еще при получении запроса в контроллере. Есть неплохая статья на тему валидации и не только.
https://struchkov.dev/blog/ru/spring-boot-validation/

SergeyLeshchev replied:
Я добавил в проект валидацию через аннотации из библиотеки и создал одну пользовательскую аннотации и валидатор для практики. Остальные поля проверяются в методе, чтобы были разные варианты решения задачи. Как лучше делать в будущем, добавлять больше пользовательских аннотаций потому что они менее бросаются в глаза? Или пользоваться аннотациями из библиотеки, а если невозможно, то валидацию делать в методе, потому что при чтении кода изучение пользовательских аннотаций и валидаторов это дополнительное неудобство?

Ivanov-s-tmn replied:
Конечно, всегда лучше использовать имеющиеся аннотации, а свои создавать только при необходимости

Ivanov-s-tmn commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,74 @@
+        film.setReleaseDate(LocalDate.parse("2000-01-01"));
+        film.setDuration(Duration.ofSeconds(-1));
+        assertThrows(ValidationException.class, () -> filmController.updateFilm(film),
+                "Валидация пропустила отрицательную продолжительность фильма");
```
null'ы тоже нужно покрыть тестами, это тоже недопустимые значения

Лучше это все вынести по разным тестам, так как много проверок в одном тесте не есть гуд, мы можем упасть и на первой проверке, и так и не узнать работает ли тест дальше

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,78 @@
+    // добавление фильма
+    @PostMapping
+    public Film addFilm(@RequestBody Film film) {
+        log.info("Вход в метод добавления нового фильма");
```
Логи стоит делать более информативными, так мы просто пишем, что произошел вход в метод, а в какой метод, контроллера или сервиса, и какой пришел запрос тоже не понятно. Такие логи не очень то полезны разработчику.\
Как пример лога:
`log.info("addFilm() - получен запрос на добавление фильма {}", film);`
Или просто 
`log.info("Создание фильма {}", film);`

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,78 @@
+        }
+
+        // дата релиза — не раньше 28 декабря 1895 года
+        if (film.getReleaseDate().isBefore(LocalDate.parse("1895-12-28"))) {
```
Дату выпуска первого фильма лучше вынести в константу, чтобы она не была магическим числом

SergeyLeshchev committed:
Выполнить правки №1 к ТЗ №10

1. Убрать ненужные аннотации и добавить нужные
2. Добавить пользовательскую аннотацию и валидатор
3. Убрать ненужную реализацию slf4j из pom
4. Добавить генерацию конструктора @AllArgsConstructor
5. Добавить тесты на проверку null в некоторых полях
6. Распределить тесты на большее количество методов
7. Сделать логи более информативными
8. Сделать константу дня рождения фильма

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -13,30 +16,18 @@
     private LocalDate releaseDate;
-    @NonNull
+    // продолжительность фильма должна быть положительным числом
+    @PositiveDuration
```
Есть базовая аннотация `@Positive`, так же можно использовать, в таком случае еще нужно добавить `@NotNull`, так как `@Positive` пропустит null как валидное

В данном случае вообще лучше использовать int и указывать время в минутах, `Duration` достаточно сложный тип и представляется в виде секунд и наносекунд, тогда для примитивного типа не нужна `@NotNull` аннотация, и можно использовать одну `@Positive` или `@Min(1)`

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -13,30 +16,18 @@
+    @NotBlank
     private String name;
-    @NonNull
+    @Size(max = 200)
```
Тут еще нужен `@NotBlank`, `@Size(max = 200)` пропустит null и пустую строку, так как нет минимального значения, но даже в этом случае будет пропущен null

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -1,32 +1,24 @@
     private String login;
     private String name;
-    @NonNull
+    @Past
```
Ссылочные типы нужно еще проверять на null, `@Past` опять же пропустит null

SergeyLeshchev committed:
Выполнить правки №2 к ТЗ №10

1. Сделать поле duration примитивом, поставить стандартную аннотацию
2. Добавить аннотации NotNull к некоторым полям

SergeyLeshchev committed:
Выполнить правки №2 к ТЗ №10

1. Сделать поле duration примитивом, поставить стандартную аннотацию
2. Добавить аннотации NotNull к некоторым полям
